src/OpenDebugAD7/Tracepoint.cs

```diff
@@ -0,0 +1,509 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.VisualStudio.Debugger.Interop;
+using Microsoft.VisualStudio.Debugger.Interop.DAP;
+using Microsoft.VisualStudio.Shared.VSCodeDebugProtocol.Messages;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace OpenDebugAD7
+{
+    internal class Tracepoint
+    {
+        // LogMessage after it has been Parsed()
+        public string LogMessage { get; set; }
+
+        private readonly IDictionary<int, string> m_indexToExpressions;
```

Comment:
are 'Expressions' the tokens that you have to evaluate separately?

Reply:
yes, it is a map of the index in LogMessage to insert the evaluated expression such as `$CALLSTACK` or `{1 + 5}`